src/container/srv_cli.c

```diff
@@ -121,6 +121,7 @@ dsc_cont_open(daos_handle_t poh, uuid_t cont_uuid, uuid_t coh_uuid,
 
 	dc_cont_hdl_link(cont);
 	dc_cont2hdl(cont, coh);
+	dc_cont_put(cont);
```

Comment:
I am not sure I followed your idea. Please check cont_open_complete(), it also takes one reference at L675 via dc_cont2hdl().

Reply:
I'm not overly familiar with this code, I was simply tracing the debug logs to see what was taken when.

Looking at cont_open_complete() however it takes a ref but then drops it at L695 again so that function will leave the ref count unchanged on the normal code-path if I've read it correctly.  There might be a secondary but there if for example the dc_cont_props_init() call failed but that's a different issue.

Reply:
As my understand, both dsc_cont_open() and cont_open_complete() hold three references: one is when dc_cont_alloc(), the other two are via dc_cont_hdl_link() and dc_cont2hdl(), then they both drop one reference via dc_cont_put(). I may missed something, you can add some debug information to check related reference.

Reply:
Without this patch there are resources leaked (the D_ALLOC_PTR from dc_cont_alloc()), I traced the reference counting though, and applying this change makes the reference count drop to zero and means the memory/resource is freed.  Without this patch the reference count stays at 1.  I'm not saying this patch is correct, but what I'm saying is that somewhere on this code-path a reference is taken but not freed.

The same issue exists for pools, however there two references are leaked rather than one.

The purpose of this PR is not to fix the issues or to land it, but to help identify areas to look at so that we can find the issue.

Reply:
Understood, then we need to find out the proper point to release the reference, or we may have to adjust related reference rules.

Reply:
I've added some more debugging to here, with this change applied (which I'm not saying is correct - I'm just looking to create discussion) then we see the following:

On the exit of dcs_cont_open() then the ref count is 1.  I hadn't previously seen the dc_cont_put() on L130, and on the entry of dcs_cont_close() the ref count is 2 and it drops two refs to so frees the memory.

What would you expect these values to be?  Obviously there are many places we call this code from and we're only testing some of them, so it may be that the caller is leaking the reference in this case and the fix appears to work because we're only testing one caller with log checking enabled.